beacon_node/network/src/subnet_service/mod.rs

```diff
@@ -17,12 +32,640 @@ pub enum SubnetServiceMessage {
+        }
+
+        // Build the list of known permanent subscriptions, so that we know not to subscribe or
+        // discover them.
+        let mut permanent_attestation_subscriptions = HashSet::default();
+        if config.subscribe_all_subnets {
+            // We are subscribed to all subnets, set all the bits to true.
+            for index in 0..beacon_chain.spec.attestation_subnet_count {
+                permanent_attestation_subscriptions
+                    .insert(Subnet::Attestation(SubnetId::from(index)));
+            }
+        } else {
+            // Not subscribed to all subnets, so just calculate the required subnets from the
+            for subnet_id in
+                SubnetId::compute_attestation_subnets(node_id.raw(), &beacon_chain.spec)
+            {
+                permanent_attestation_subscriptions.insert(Subnet::Attestation(subnet_id));
+            }
+        }
+
+        // Set up the sync committee subscriptions
+        let spec = &beacon_chain.spec;
+        let epoch_duration_secs =
+            beacon_chain.slot_clock.slot_duration().as_secs() * T::EthSpec::slots_per_epoch();
+        let default_sync_committee_duration = Duration::from_secs(
+            epoch_duration_secs.saturating_mul(spec.epochs_per_sync_committee_period.as_u64()),
+        );
+
+        let track_validators = !config.import_all_attestations;
+        let aggregate_validators_on_subnet =
+            track_validators.then(|| HashSetDelay::new(slot_duration));
+
+        let mut events = VecDeque::with_capacity(10);
+
+        // Queue discovery queries for the permanent attestation subnets
+        if !config.disable_discovery {
+            events.push_back(SubnetServiceMessage::DiscoverPeers(
+                permanent_attestation_subscriptions
+                    .iter()
+                    .cloned()
+                    .map(|subnet| SubnetDiscovery {
+                        subnet,
+                        min_ttl: None,
+                    })
+                    .collect(),
+            ));
+        }
+
+        // Pre-populate the events with permanent subscriptions
+        for subnet in permanent_attestation_subscriptions.iter() {
+            events.push_back(SubnetServiceMessage::Subscribe(*subnet));
+            events.push_back(SubnetServiceMessage::EnrAdd(*subnet));
+        }
+
+        SubnetService {
+            events,
+            beacon_chain,
+            subscriptions: HashSetDelay::new(default_sync_committee_duration),
+            permanent_attestation_subscriptions,
+            scheduled_subscriptions: HashSetDelay::default(),
+            aggregate_validators_on_subnet,
+            waker: None,
+            discovery_disabled: config.disable_discovery,
+            subscribe_all_subnets: config.subscribe_all_subnets,
+            proposer_only: config.proposer_only,
+            log,
+        }
+    }
+
+    /// Return count of all currently subscribed subnets (long-lived **and** short-lived).
```

Comment:
short-lived subscriptions only right?

```suggestion
    /// Return count of all currently subscribed subnets (short-lived).
```